libkbfs/journal_server.go

```diff
@@ -104,26 +110,37 @@ func makeJournalServer(
 	return &jServer
 }
 
-func (j *JournalServer) getDirLocked() string {
+func (j *JournalServer) rootPath() string {
+	return filepath.Join(j.dir, "v1")
+}
+
+func (j *JournalServer) tlfJournalPathLocked(tlfID TlfID) string {
 	if j.currentVerifyingKey == (VerifyingKey{}) {
 		panic("currentVerifyingKey is zero")
 	}
 
-	// Device IDs and verifying keys are globally unique, so no
-	// need to have the uid in the path. Furthermore, everything
-	// after the first two bytes (four characters) are randomly
-	// generated, so taking the first 36 characters gives ~1/2^64
-	// collision probability.
+	// We need to generate a unique path for each (UID, device,
+	// TLF) tuple. Verifying keys (which are unique to a device)
+	// are globally unique, so no need to have the uid in the
+	// path. Furthermore, everything after the first two bytes
+	// (four characters) is randomly generated, so taking the
+	// first 36 characters of the verifying key gives us 16 random
+	// bytes (since the first two bytes encode version/type) or
+	// 128 random bits, which means that the expected number of
+	// devices generated before getting a collision in the first
+	// part of the path is 2^64 (see
+	// https://en.wikipedia.org/wiki/Birthday_problem#Cast_as_a_collision_problem
+	// ).
 	//
-	// TODO: Write out a file with the full string.
-	shortID := j.currentVerifyingKey.String()[0:36]
-	return filepath.Join(j.dir, "v1", shortID)
-}
-
-func (j *JournalServer) getDir() string {
-	j.lock.RLock()
-	defer j.lock.RUnlock()
-	return j.getDirLocked()
+	// By similar reasoning, for a single device, taking the first
+	// 16 characters of the TLF ID gives us 64 random bits, which
+	// means that the expected number of TLFs associated to that
+	// device before getting a collision in the second part of the
+	// path is 2^32.
+	shortDeviceIDStr := j.currentVerifyingKey.String()[:36]
```

Comment:
Unfortunately, I had to bump this back up to 36 characters, bumping up the total number of characters added from a nice 100 to 116. Taking 8 bytes from the device ID and 8 bytes from the TLF ID only works if we assume that we're picking devices and TLFs independently and mashing them together, which isn't what happens; the distribution of TLF IDs isn't independent from the distribution of device IDs.

This way, we still get 16 bytes of randomness from the device ID, and we assume that for a single device ID, it won't be associated with anything close to 2^32 TLFs.